File: example/shared_memory/queue_caller.py
"""Two bounded queues: synthetic RGB frames -> C detection -> slot returns."""
from __future__ import annotations

import argparse
from collections import deque
import json
import math
from multiprocessing.shared_memory import SharedMemory
from pathlib import Path
import threading
import time

from synurang import ModuleHost
from frame_queue_client import FrameQueueClient
from frame_queue_lite import Frame, FrameEvent, QueueConfig, QueuePolicy, QueueRequest

WIDTH, HEIGHT, BOX, GUARD = 96, 64, 12, 32
PIXELS = WIDTH * HEIGHT * 3
STRIDE = PIXELS + GUARD


def target(frame_id: int) -> tuple[int, int]:
    return (frame_id * 3) % (WIDTH - BOX + 1), (frame_id * 2) % (HEIGHT - BOX + 1)


def has_defect(frame_id: int, scenario: str) -> bool:
    # Adjacent defect frames avoid hiding skipped inspections through a lucky
    # alignment between a periodic defect and the worker's sampling cadence.
    return scenario == "inspection" and frame_id % 7 in (6, 0)


def produce(region: SharedMemory, slot: int, frame_id: int, scenario: str) -> None:
    start = slot * STRIDE
    # Generate directly into the pool, without an image-sized staging object.
    for i in range(PIXELS):
        region.buf[start + i] = 20
    for i in range(PIXELS, STRIDE):
        region.buf[start + i] = 0xCD
    x, y = target(frame_id)
    for row in range(y, y + BOX):
        for col in range(x, x + BOX):
            region.buf[start + (row * WIDTH + col) * 3] = 220
    if has_defect(frame_id, scenario):
        for row in range(y + 5, y + 7):
            for col in range(x + 5, x + 7):
                p = start + (row * WIDTH + col) * 3
                region.buf[p], region.buf[p + 2] = 20, 240


def verify(region: SharedMemory, result, scenario: str) -> None:
    processed = result.event == FrameEvent.PROCESSED
    x, y = target(result.frame_id)
    defect = has_defect(result.frame_id, scenario)
    if processed:
        assert result.detected and (result.x, result.y, result.width, result.height) == (x, y, BOX, BOX)
        assert result.defect == defect
    start = result.slot * STRIDE
    for row in range(HEIGHT):
        for col in range(WIDTH):
            red, green, blue = 20, 20, 20
            if x <= col < x + BOX and y <= row < y + BOX:
                red = 220
                if defect and x + 5 <= col < x + 7 and y + 5 <= row < y + 7:
                    red, blue = 20, 240
                if processed and (col in (x, x + BOX - 1) or row in (y, y + BOX - 1)):
                    red, green, blue = (255 if defect else 0), 255, 0
            p = start + (row * WIDTH + col) * 3
            assert region.buf[p] == red and region.buf[p + 1] == green and region.buf[p + 2] == blue
    assert all(region.buf[start + i] == 0xCD for i in range(PIXELS, STRIDE))


def pool_config(region, *, policy="fifo", slots=16, input_capacity=8, output_capacity=4,
                batch_size=4, batch_wait_ms=50, work_ms=40):
    return QueueConfig(name="/" + region.name.lstrip("/"), width=WIDTH, height=HEIGHT,
                       slot_stride=STRIDE, slot_count=slots, policy=QueuePolicy[policy.upper()],
                       input_capacity=input_capacity, output_capacity=output_capacity,
                       batch_size=batch_size, batch_wait_ms=batch_wait_ms, simulate_work_ms=work_ms)


def percentile(values, fraction):
    return sorted(values)[max(0, math.ceil(len(values) * fraction) - 1)] if values else 0


def run(build: Path, *, policy="fifo", scenario="tracking", frames=60, fps=60.0,
        work_ms=40, slots=16, input_capacity=8, output_capacity=4, batch_size=4,
        batch_wait_ms=50, consumer_delay_ms=0, output: Path | None = None, quiet=False):
    if not (1 <= frames <= 10000 and 1 <= slots <= 64 and math.isfinite(fps)
            and (fps == 0 or .1 <= fps <= 10000) and 0 <= work_ms <= 1000
            and 0 <= consumer_delay_ms <= 1000):
        raise ValueError("frames=1..10000, slots=1..64, fps=0 or 0.1..10000, delays=0..1000 required")
    if output:
        output.mkdir(parents=True, exist_ok=True)
    region = SharedMemory(create=True, size=slots * STRIDE)
    condition, stop = threading.Condition(), threading.Event()
    available, outstanding, errors = deque(range(slots)), {}, []
    events, latencies = [], []
    started = time.monotonic_ns()
    summary = None
    try:
        # Host teardown precedes pool teardown, including any failed send/recv.
        with ModuleHost.load(build / "backend.so", loader=build / "libsynurang_module_host.so") as host:
            timeout = max(10, frames * ((1 / fps if fps else 0) + (work_ms + consumer_delay_ms + 100) / 1000) + 5)
            with FrameQueueClient(host).run(timeout=timeout) as call:
                config = pool_config(region, policy=policy, slots=slots, input_capacity=input_capacity,
                                     output_capacity=output_capacity, batch_size=batch_size,
                                     batch_wait_ms=batch_wait_ms, work_ms=work_ms)
                call.send(QueueRequest(config=config))
                ready = call.recv()
                assert ready is not None and ready.event == FrameEvent.READY

                def send_frames():
                    next_capture = time.monotonic()
                    try:
                        for frame_id in range(1, frames + 1):
                            if stop.wait(max(0, next_capture - time.monotonic())):
                                return
                            with condition:
                                condition.wait_for(lambda: available or stop.is_set())
                                if stop.is_set():
                                    return
                                slot = available.popleft()
                            captured = time.monotonic_ns()
                            next_capture = time.monotonic() + (1 / fps if fps else 0)
                            produce(region, slot, frame_id, scenario)
                            with condition:
                                outstanding[frame_id] = (slot, captured)
                            call.send(QueueRequest(frame=Frame(slot=slot, frame_id=frame_id, captured_ns=captured)))
                        call.half_close()  # Flush the final, possibly partial batch.
                    except BaseException as error:
                        errors.append(error)
                        stop.set()
                        call.cancel()
                        with condition:
                            condition.notify_all()

                producer = threading.Thread(target=send_frames, name="frame-producer")
                producer.start()
                try:
                    for result in call:
                        received = time.monotonic_ns()
                        if result.event == FrameEvent.SUMMARY:
                            assert summary is None
                            summary = result
                            continue
                        assert summary is None and result.event in (FrameEvent.PROCESSED, FrameEvent.DROPPED)
                        with condition:
                            assert outstanding[result.frame_id] == (result.slot, result.captured_ns)
                        verify(region, result, scenario)
                        if result.event == FrameEvent.PROCESSED:
                            assert result.captured_ns <= result.started_ns <= result.completed_ns <= received
                            latencies.append((received - result.captured_ns) / 1e6)
                            if output and result.frame_id == frames:
                                with (output / f"python_{policy}_{scenario}.ppm").open("wb") as image:
                                    image.write(f"P6\n{WIDTH} {HEIGHT}\n255\n".encode())
                                    view = region.buf[result.slot * STRIDE:result.slot * STRIDE + PIXELS]
                                    try:
                                        image.write(view)
                                    finally:
                                        view.release()
                        events.append(dict(id=result.frame_id, slot=result.slot, status=FrameEvent(result.event).name,
                                           captured_ms=(result.captured_ns - started) / 1e6,
                                           received_ms=(received - started) / 1e6,
                                           completed_ms=(result.completed_ns - started) / 1e6,
                                           batch_id=result.batch_id, batch_size=result.batch_size,
                                           x=result.x, y=result.y, defect=bool(result.defect)))
                        with condition:
                            del outstanding[result.frame_id]
                            available.append(result.slot)
                            condition.notify_all()
                        if consumer_delay_ms:
                            time.sleep(consumer_delay_ms / 1000)
                    producer.join()
                    if errors:
                        raise errors[0]
                    assert summary is not None and not outstanding and len(available) == slots
                    assert len(events) == frames and summary.processed + summary.dropped == frames
                    assert summary.processed == len(latencies)
                    if policy != "latest":
                        assert summary.dropped == 0
                    assert [e["id"] for e in events] == list(range(1, frames + 1))
                finally:
                    stop.set()
                    with condition:
                        condition.notify_all()
                    call.cancel()
                    producer.join()
    finally:
        region.close()
        region.unlink()
    report = dict(caller="Python", policy=policy, scenario=scenario, frames=frames,
                  processed=summary.processed, dropped=summary.dropped, batches=summary.batches,
                  max_batch_size=max(e["batch_size"] for e in events),
                  detected_defects=sum(e["defect"] for e in events),
                  missed_defects=sum(e["status"] == "DROPPED" and has_defect(e["id"], scenario) for e in events),
                  input_high_water=summary.input_high_water, output_high_water=summary.output_high_water,
                  outstanding=0, latency_p50_ms=percentile(latencies, .5),
                  latency_p95_ms=percentile(latencies, .95), events=events)
    if output:
        (output / f"python_{policy}_{scenario}.json").write_text(json.dumps(report, indent=2) + "\n")
    if not quiet:
        print(f"Python {policy:6} {scenario:10}: processed={report['processed']} dropped={report['dropped']} "
              f"latency p50/p95={report['latency_p50_ms']:.1f}/{report['latency_p95_ms']:.1f} ms "
              f"batch<={report['max_batch_size']} defects={report['detected_defects']} "
              f"missed={report['missed_defects']} outstanding=0")
    return report


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("build", type=Path)
    parser.add_argument("--policy", choices=("fifo", "latest", "batch"), default="fifo")
    parser.add_argument("--scenario", choices=("tracking", "inspection"), default="tracking")
    parser.add_argument("--frames", type=int, default=60)
    parser.add_argument("--fps", type=float, default=60, help="Synthetic capture rate; 0 sends as fast as pool permits")
    parser.add_argument("--work-ms", type=int, default=40, help="Simulated per-frame wait; 0 disables it")
    parser.add_argument("--slots", type=int, default=16)
    parser.add_argument("--input-capacity", type=int, default=8)
    parser.add_argument("--output-capacity", type=int, default=4)
    parser.add_argument("--batch-size", type=int, default=4)
    parser.add_argument("--batch-wait-ms", type=int, default=50)
    parser.add_argument("--consumer-delay-ms", type=int, default=0)
    parser.add_argument("--output", type=Path, help="Save metadata JSON and the final annotated PPM")
    arguments = vars(parser.parse_args())
    arguments["build"] = arguments["build"].resolve()
    run(**arguments)

File: example/shared_memory/queue_caller.cpp
#include "frame_queue_lite.h"
#include "shared_region.hpp"
#include <synurang/module_host.hpp>

#include <algorithm>
#include <cmath>
#include <condition_variable>
#include <ctime>
#include <deque>
#include <exception>
#include <filesystem>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <memory>
#include <mutex>
#include <thread>
#include <vector>

using Clock = std::chrono::steady_clock;
using Result = SynurangExampleShmFrameResult;
constexpr std::uint32_t width = 96, height = 64, box = 12, guard = 32;
constexpr std::size_t pixels = width * height * 3, stride = pixels + guard;
constexpr auto processed = SYNURANG_EXAMPLE_SHM_FRAME_EVENT_PROCESSED;
constexpr auto dropped = SYNURANG_EXAMPLE_SHM_FRAME_EVENT_DROPPED;

static void require(bool value, const char* message) {
    if (!value) throw std::runtime_error(message);
}
static std::uint64_t now_ns() {
    // Use the same POSIX monotonic clock as C and Python for timestamp checks.
    timespec time{};
    clock_gettime(CLOCK_MONOTONIC, &time);
    return static_cast<std::uint64_t>(time.tv_sec) * 1000000000 + time.tv_nsec;
}
struct Options {
    std::filesystem::path build, output;
    std::string policy = "fifo", scenario = "tracking";
    int frames = 60, work_ms = 40, slots = 16, input_capacity = 8, output_capacity = 4;
    int batch_size = 4, batch_wait_ms = 50, consumer_delay_ms = 0;
    double fps = 60;
};
static bool defect_for(std::uint64_t id, const Options& options) {
    return options.scenario == "inspection" && (id % 7 == 6 || id % 7 == 0);
}
static void produce(const SharedRegion& region, std::uint32_t slot, std::uint64_t id, const Options& options) {
    auto* data = region.data() + slot * stride;
    std::fill_n(data, pixels, 20);
    std::fill_n(data + pixels, guard, 0xCD);
    const auto x = id * 3 % (width - box + 1), y = id * 2 % (height - box + 1);
    for (auto row = y; row < y + box; ++row)
        for (auto col = x; col < x + box; ++col) data[(row * width + col) * 3] = 220;
    if (defect_for(id, options)) {
        for (auto row = y + 5; row < y + 7; ++row) {
            for (auto col = x + 5; col < x + 7; ++col) {
                auto* p = data + (row * width + col) * 3;
                p[0] = 20; p[2] = 240;
            }
        }
    }
}
static void verify(const SharedRegion& region, const Result& result, const Options& options) {
    const auto x = result.field_frame_id * 3 % (width - box + 1);
    const auto y = result.field_frame_id * 2 % (height - box + 1);
    const bool annotated = result.field_event == processed;
    const bool defect = defect_for(result.field_frame_id, options);
    if (annotated) {
        require(result.field_detected && result.field_x == x && result.field_y == y &&
                result.field_width == box && result.field_height == box, "Incorrect bounding box");
        require(static_cast<bool>(result.field_defect) == defect, "Incorrect defect detection");
    }
    const auto* data = region.data() + result.field_slot * stride;
    for (std::uint32_t row = 0; row < height; ++row) {
        for (std::uint32_t col = 0; col < width; ++col) {
            int red = 20, green = 20, blue = 20;
            if (col >= x && col < x + box && row >= y && row < y + box) {
                red = 220;
                if (defect && col >= x + 5 && col < x + 7 && row >= y + 5 && row < y + 7) {
                    red = 20; blue = 240;
                }
                if (annotated && (col == x || col == x + box - 1 || row == y || row == y + box - 1)) {
                    red = defect ? 255 : 0; green = 255; blue = 0;
                }
            }
            const auto* p = data + (row * width + col) * 3;
            require(p[0] == red && p[1] == green && p[2] == blue, "Incorrect shared pixel or premature slot reuse");
        }
    }
    for (std::size_t i = pixels; i < stride; ++i) require(data[i] == 0xCD, "Slot guard modified");
}
static synurang::ModuleBytes encode(const SynurangExampleShmQueueRequest& request) {
    std::uint8_t* bytes = nullptr;
    std::size_t size = 0;
    auto status = synurang_example_shm_queue_request_encode(&request, &bytes, &size);
    std::unique_ptr<std::uint8_t, decltype(&std::free)> owned(bytes, &std::free);
    require(status == SYNURANG_LITE_OK, "Cannot encode queue descriptor");
    return {bytes, bytes + size};
}
static Result decode(const synurang::ModuleBytes& bytes) {
    Result result;
    synurang_example_shm_frame_result_init(&result);
    require(synurang_example_shm_frame_result_decode(&result, bytes.data(), bytes.size()) == SYNURANG_LITE_OK,
            "Cannot decode frame result");
    return result; // Scalar-only message, with no owned allocations.
}
static void configure(const synurang::ModuleCall& call, const SharedRegion& region, const Options& o) {
    SynurangExampleShmQueueConfig config;
    SynurangExampleShmQueueRequest request;
    synurang_example_shm_queue_config_init(&config);
    synurang_example_shm_queue_request_init(&request);
    config.field_name.data = reinterpret_cast<std::uint8_t*>(const_cast<char*>(region.name().data()));
    config.field_name.len = region.name().size();
    config.field_width = width; config.field_height = height;
    config.field_slot_stride = stride; config.field_slot_count = o.slots;
    config.field_policy = o.policy == "fifo" ? SYNURANG_EXAMPLE_SHM_QUEUE_POLICY_FIFO :
                          o.policy == "latest" ? SYNURANG_EXAMPLE_SHM_QUEUE_POLICY_LATEST :
                                                 SYNURANG_EXAMPLE_SHM_QUEUE_POLICY_BATCH;
    config.field_input_capacity = o.input_capacity; config.field_output_capacity = o.output_capacity;
    config.field_batch_size = o.batch_size; config.field_batch_wait_ms = o.batch_wait_ms;
    config.field_simulate_work_ms = o.work_ms;
    request.which_value = 1; request.field_config = &config;
    call.send(encode(request)); // Stack/name pointers are borrowed by encode, never freed as messages.
    auto ready = call.receive();
    require(ready && decode(*ready).field_event == SYNURANG_EXAMPLE_SHM_FRAME_EVENT_READY, "Missing READY");
}
struct Event { Result result; std::uint64_t received; };
static double percentile(std::vector<double> values, double fraction) {
    if (values.empty()) return 0;
    std::sort(values.begin(), values.end());
    return values[static_cast<std::size_t>(std::ceil(values.size() * fraction)) - 1];
}

static void run(const Options& o) {
    SharedRegion region(static_cast<std::size_t>(o.slots) * stride);
    auto host = synurang::ModuleHost::load((o.build / "backend.so").string());
    synurang::ModuleCallOptions call_options;
    const double timeout_ms = std::max(10000.0, o.frames * ((o.fps ? 1000 / o.fps : 0) +
                                                         o.work_ms + o.consumer_delay_ms + 100) + 5000);
    call_options.timeout = std::chrono::milliseconds(static_cast<long long>(std::ceil(timeout_ms)));
    auto call = host.open("/synurang.example.shm.FrameQueue/Run", true, true, call_options);
    const auto started = now_ns();
    configure(call, region, o);
    if (!o.output.empty()) std::filesystem::create_directories(o.output);
    std::mutex mutex;
    std::condition_variable changed;
    bool stop = false;
    std::deque<std::uint32_t> available;
    struct Owner { std::uint64_t id = 0, captured = 0; };
    std::vector<Owner> owners(o.slots);
    std::vector<Event> events;
    std::vector<double> latencies;
    std::exception_ptr producer_error;
    Result summary{};
    for (int i = 0; i < o.slots; ++i) available.push_back(i);
    std::thread producer([&] {
        try {
            auto next_capture = Clock::now();
            for (int id = 1; id <= o.frames; ++id) {
                std::uint32_t slot;
                {
                    std::unique_lock<std::mutex> lock(mutex);
                    if (changed.wait_until(lock, next_capture, [&] { return stop; })) return;
                    changed.wait(lock, [&] { return stop || !available.empty(); });
                    if (stop) return;
                    slot = available.front(); available.pop_front();
                }
                const auto captured = now_ns();
                next_capture = Clock::now() + std::chrono::duration_cast<Clock::duration>(
                    std::chrono::duration<double>(o.fps ? 1 / o.fps : 0));
                produce(region, slot, id, o);
                {
                    std::lock_guard<std::mutex> lock(mutex);
                    owners[slot] = {static_cast<std::uint64_t>(id), captured};
                }
                SynurangExampleShmFrame frame;
                SynurangExampleShmQueueRequest request;
                synurang_example_shm_frame_init(&frame);
                synurang_example_shm_queue_request_init(&request);
                frame.field_slot = slot; frame.field_frame_id = id; frame.field_captured_ns = captured;
                request.which_value = 2; request.field_frame = &frame;
                call.send(encode(request));
            }
            call.half_close();
        } catch (...) {
            producer_error = std::current_exception();
            { std::lock_guard<std::mutex> lock(mutex); stop = true; }
            changed.notify_all();
            call.cancel();
        }
    });
    try {
        while (auto bytes = call.receive()) {
            const auto received = now_ns();
            const auto result = decode(*bytes);
            if (result.field_event == SYNURANG_EXAMPLE_SHM_FRAME_EVENT_SUMMARY) {
                require(!summary.field_event, "Duplicate SUMMARY");
                summary = result;
                continue;
            }
            require(!summary.field_event && (result.field_event == processed || result.field_event == dropped), "Invalid event");
            require(result.field_slot < owners.size() && result.field_frame_id == events.size() + 1, "Duplicate or out-of-order return");
            {
                std::lock_guard<std::mutex> lock(mutex);
                require(owners[result.field_slot].id == result.field_frame_id &&
                        owners[result.field_slot].captured == result.field_captured_ns, "Invalid slot return");
            }
            verify(region, result, o);
            if (result.field_event == processed) {
                require(result.field_captured_ns <= result.field_started_ns && result.field_started_ns <= result.field_completed_ns &&
                        result.field_completed_ns <= received, "Invalid timestamps");
                latencies.push_back(static_cast<double>(received - result.field_captured_ns) / 1e6);
                if (!o.output.empty() && result.field_frame_id == static_cast<std::uint64_t>(o.frames)) {
                    std::ofstream image(o.output / ("cpp_" + o.policy + "_" + o.scenario + ".ppm"), std::ios::binary);
                    image << "P6\n" << width << ' ' << height << "\n255\n";
                    image.write(reinterpret_cast<const char*>(region.data() + result.field_slot * stride), pixels);
                    require(static_cast<bool>(image), "Cannot save annotated PPM");
                }
            }
            events.push_back({result, received});
            {
                std::lock_guard<std::mutex> lock(mutex);
                owners[result.field_slot] = {};
                available.push_back(result.field_slot);
            }
            changed.notify_all();
            if (o.consumer_delay_ms) std::this_thread::sleep_for(std::chrono::milliseconds(o.consumer_delay_ms));
        }
        producer.join();
        if (producer_error) std::rethrow_exception(producer_error);
        require(summary.field_event && events.size() == static_cast<std::size_t>(o.frames) &&
                summary.field_processed + summary.field_dropped == events.size() &&
                summary.field_processed == latencies.size() && available.size() == owners.size(), "Missing slot returns");
        if (o.policy != "latest") require(summary.field_dropped == 0, "Lossless policy dropped a frame");
        for (const auto& owner : owners) require(owner.id == 0, "Outstanding buffer at shutdown");
    } catch (...) {
        { std::lock_guard<std::mutex> lock(mutex); stop = true; }
        changed.notify_all();
        call.cancel();
        if (producer.joinable()) producer.join();
        host.close(); // Worker joins before region unmaps, even without ACKs.
        throw;
    }
    host.close();
    int defects = 0, missed = 0;
    std::uint32_t max_batch = 0;
    for (const auto& event : events) {
        defects += event.result.field_defect != 0;
        missed += event.result.field_event == dropped && defect_for(event.result.field_frame_id, o);
        max_batch = std::max(max_batch, event.result.field_batch_size);
    }
    const double p50 = percentile(latencies, .5), p95 = percentile(latencies, .95);
    if (!o.output.empty()) {
        std::ofstream json(o.output / ("cpp_" + o.policy + "_" + o.scenario + ".json"));
        json << std::fixed << std::setprecision(3) << "{\"caller\":\"C++\",\"policy\":\"" << o.policy
             << "\",\"scenario\":\"" << o.scenario << "\",\"frames\":" << o.frames
             << ",\"processed\":" << summary.field_processed << ",\"dropped\":" << summary.field_dropped
             << ",\"batches\":" << summary.field_batches << ",\"max_batch_size\":" << max_batch
             << ",\"detected_defects\":" << defects << ",\"missed_defects\":" << missed
             << ",\"input_high_water\":" << summary.field_input_high_water
             << ",\"output_high_water\":" << summary.field_output_high_water
             << ",\"outstanding\":0,\"latency_p50_ms\":" << p50 << ",\"latency_p95_ms\":" << p95 << ",\"events\":[";
        bool first = true;
        for (const auto& event : events) {
            const auto& r = event.result;
            json << (first ? "" : ",") << "{\"id\":" << r.field_frame_id << ",\"slot\":" << r.field_slot
                 << ",\"status\":\"" << (r.field_event == processed ? "PROCESSED" : "DROPPED")
                 << "\",\"captured_ms\":" << static_cast<double>(r.field_captured_ns - started) / 1e6
                 << ",\"received_ms\":" << static_cast<double>(event.received - started) / 1e6
                 << ",\"completed_ms\":" << static_cast<double>(r.field_completed_ns - started) / 1e6
                 << ",\"batch_id\":" << r.field_batch_id << ",\"batch_size\":" << r.field_batch_size
                 << ",\"x\":" << r.field_x << ",\"y\":" << r.field_y
                 << ",\"defect\":" << (r.field_defect ? "true" : "false") << '}';
            first = false;
        }
        json << "]}\n";
        require(static_cast<bool>(json), "Cannot save queue report");
    }
    std::cout << std::fixed << std::setprecision(1) << "C++ " << o.policy << ' ' << o.scenario
              << ": processed=" << summary.field_processed << " dropped=" << summary.field_dropped
              << " latency p50/p95=" << p50 << '/' << p95 << " ms batch<=" << max_batch
              << " defects=" << defects << " missed=" << missed << " outstanding=0\n";
}

int main(int argc, char** argv) {
    try {
        require(argc >= 2 && argc % 2 == 0,
                "Usage: cpp_queue_caller BUILD [--policy fifo|latest|batch] [--scenario tracking|inspection] "
                "[--frames N] [--fps N] [--work-ms N] [--slots N] [--input-capacity N] [--output-capacity N] "
                "[--batch-size N] [--batch-wait-ms N] [--consumer-delay-ms N] [--output DIR]");
        Options options;
        options.build = argv[1];
        for (int i = 2; i < argc; i += 2) {
            const std::string key = argv[i], value = argv[i + 1];
            if (key == "--policy") options.policy = value;
            else if (key == "--scenario") options.scenario = value;
            else if (key == "--output") options.output = value;
            else if (key == "--fps") options.fps = std::stod(value);
            else if (key == "--frames") options.frames = std::stoi(value);
            else if (key == "--work-ms") options.work_ms = std::stoi(value);
            else if (key == "--slots") options.slots = std::stoi(value);
            else if (key == "--input-capacity") options.input_capacity = std::stoi(value);
            else if (key == "--output-capacity") options.output_capacity = std::stoi(value);
            else if (key == "--batch-size") options.batch_size = std::stoi(value);
            else if (key == "--batch-wait-ms") options.batch_wait_ms = std::stoi(value);
            else if (key == "--consumer-delay-ms") options.consumer_delay_ms = std::stoi(value);
            else throw std::runtime_error("Unknown option: " + key);
        }
        require(options.policy == "fifo" || options.policy == "latest" || options.policy == "batch", "Invalid policy");
        require(options.scenario == "tracking" || options.scenario == "inspection", "Invalid scenario");
        require(options.frames >= 1 && options.frames <= 10000 && options.slots >= 1 && options.slots <= 64 &&
                std::isfinite(options.fps) && (options.fps == 0 || (options.fps >= .1 && options.fps <= 10000)) &&
                options.work_ms >= 0 && options.work_ms <= 1000 &&
                options.consumer_delay_ms >= 0 && options.consumer_delay_ms <= 1000, "Invalid rate, count or delay");
        run(options);
    } catch (const std::exception& error) {
        std::cerr << error.what() << '\n';
        return EXIT_FAILURE;
    }
}
